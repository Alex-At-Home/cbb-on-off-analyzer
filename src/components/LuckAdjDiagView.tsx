// React imports:
import React, { useState } from 'react';

// Next imports:
import { NextPage } from 'next';

import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Tooltip from 'react-bootstrap/Tooltip';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';

// Utils
import { StatsUtils } from "../utils/StatsUtils";
import { CommonTableDefs } from "../utils/CommonTableDefs";
import { CbbColors } from "../utils/CbbColors";

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable";
import { TeamStatsModel } from '../components/TeamStatsTable';
import { RosterStatsModel } from '../components/RosterStatsTable';

type Props = {
  teamStats: TeamStatsModel,
  rosterStats: RosterStatsModel
};
const LuckAdjDiagView: React.FunctionComponent<Props> = ({teamStats, rosterStats}) => {

  const topRef = React.createRef<HTMLDivElement>();

  return <span ref={topRef}>
      <span>
        <b>Luck Adjustment diagnostics [TODO]</b>
      </span>
      <ul>
        <li>Defense</li>
        <ul>
          <li>GLOBAL POSS {(teamStats?.global?.def_poss?.value || 0).toFixed(0)}</li>
          <li>GLOBAL 3P SOS {(teamStats?.global?.def_3p_opp?.value || 0).toFixed(1)}</li>
          <li>BASE 3P SOS {(teamStats?.baseline?.def_3p_opp?.value || 0).toFixed(1)}</li>
          <li>ON 3P SOS {(teamStats?.on?.def_3p_opp?.value || 0).toFixed(1)}</li>
          <li>OFF 3P SOS {(teamStats?.off?.def_3p_opp?.value || 0).toFixed(1)}</li>
        </ul>
      </ul>

      (<a href="#" onClick={(event) => {
        event.preventDefault();
        if (topRef.current) {
          topRef.current.scrollIntoView({
            behavior: 'smooth',
            block: 'start',
          });
        }
      }}>
        Scroll back to start of positional diagnostics
      </a>)
    </span>;
};
export default LuckAdjDiagView;
