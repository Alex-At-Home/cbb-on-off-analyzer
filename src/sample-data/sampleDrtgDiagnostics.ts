export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
