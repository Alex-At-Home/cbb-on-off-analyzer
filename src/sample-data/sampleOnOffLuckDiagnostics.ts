
/** Manually built by copying the snapshot from src/utils/stats/__tests__/LuckUtils.test.ts */
export const sampleOffOnOffLuckDiagnostics =
{
   "avgEff": 100,
   "samplePoss": 1602,
   "sample3P": 0.31958762886597936,
   "sample3PA": 582,
   "base3PA": 765,
   "player3PInfo": {
      "Cowan, Anthony": {
         "shot_info_ast_3pm": 35,
         "shot_info_early_3pa": 23,
         "shot_info_unast_3pm": 13,
         "shot_info_unknown_3pM": 103,
         "shot_info_total": 174,
         "base3P": 0.3218390804597701,
         "unassisted3P": 0.2780890804597701,
         "assisted3P": 0.3380890804597701,
         "expected3P": 0.3218390804597701
      },
      "Wiggins, Aaron": {
         "shot_info_ast_3pm": 49,
         "shot_info_early_3pa": 47,
         "shot_info_unast_3pm": 0,
         "shot_info_unknown_3pM": 71,
         "shot_info_total": 167,
         "base3P": 0.31736526946107785,
         "unassisted3P": 0.25736526946107785,
         "assisted3P": 0.31736526946107785,
         "expected3P": 0.31736526946107785
      }
   },
   "sampleBase3P": 0.3196480938416422,
   "regress3P": 0.31962196866284803,
   "sampleOff3PRate": 0.4279411764705882,
   "sampleOffFGA": 1360,
   "sampleOffOrb": 0.30924170616113744,
   "sampleOffEfg": 0.49779411764705883,
   "sampleOffPpp": 108.3645443196005,
   "sampleDefSos": 96.33476903870161,
   "delta3P": 0.0000343397968686765,
   "deltaOffEfg": 0.00002204311960761366,
   "deltaMissesPct": -0.000014695413071742441,
   "deltaOffPppNoOrb": 0.0037426520182714832,
   "deltaOffOrbFactor": -0.00000454441395925623,
   "deltaPtsOffMisses": -0.0004924703560545094,
   "deltaOffPpp": 0.003250181662216974,
   "deltaOffAdjEff": 0.0033738407167522694
};


/** Manually built by copying the snapshot from src/utils/stats/__tests__/LuckUtils.test.ts */
export const sampleDefOnOffLuckDiagnostics =
{
   "avgEff": 100,
   "luckPct": 0.66,
   "baseDef3P": 0.32521489971346706,
   "baseDef3PSos": 0.3298939828080229,
   "basePoss": 2081,
   "base3PSosAdj": -0.0015908882521489841,
   "sampleDef3P": 0.3614457831325301,
   "sampleDef3PSos": 0.3225421686746988,
   "samplePoss": 489,
   "sample3PSosAdj": 0.013227228915662642,
   "sampleDefEfg": 0.4733178654292343,
   "sampleDefPpp": 98.15950920245399,
   "sampleOffSos": 105.45848670756645,
   "sampleDef3PRate": 0.3851508120649652,
   "sampleDefFGA": 431,
   "sampleDefOrb": 0.2947761194029851,
   "avg3PSosAdj": 0.001228590072777041,
   "adjDef3P": 0.32377075874747585,
   "delta3P": -0.03767502438505427,
   "deltaDefEfg": -0.021765849354706525,
   "deltaDefPppNoOrb": -3.836842974183441,
   "deltaMissesPct": 0.014510566236471017,
   "deltaDefOrbFactor": 0.004295742880401803,
   "deltaPtsOffMisses": 0.40518592191060876,
   "deltaDefPpp": -3.4316570522728322,
   "deltaDefAdjEff": -3.2540359333893396
};
