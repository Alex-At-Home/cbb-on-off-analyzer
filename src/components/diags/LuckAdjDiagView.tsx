// React imports:
import React, { useState } from 'react';

// Next imports:
import { NextPage } from 'next';

import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Tooltip from 'react-bootstrap/Tooltip';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';

// Utils
import { CommonTableDefs } from "../../utils/CommonTableDefs";
import { CbbColors } from "../../utils/CbbColors";
import { LuckAdjustmentBaseline, OffLuckAdjustmentDiags, DefLuckAdjustmentDiags } from "../../utils/stats/LuckUtils";

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "../GenericTable";
import { TeamStatsModel } from '../../components/TeamStatsTable';
import { RosterStatsModel } from '../../components/RosterStatsTable';

type Props = {
  name: string,
  offLuck: OffLuckAdjustmentDiags,
  defLuck: DefLuckAdjustmentDiags,
  baseline: LuckAdjustmentBaseline,
  individualMode?: boolean,
  showHelp: boolean,
  showDetailsOverride?: boolean
};
const LuckAdjDiagView: React.FunctionComponent<Props> = ({name, offLuck, defLuck, baseline, individualMode, showHelp, showDetailsOverride}) => {

  const topRef = React.createRef<HTMLDivElement>();

  const [ showDetails, setShowDetails ] = useState(true);
  const [ show3POff, setShow3POff ] = useState(_.isNil(showDetailsOverride) ? false : showDetailsOverride);
  const [ show3PDef, setShow3PDef ] = useState(_.isNil(showDetailsOverride) ? false : showDetailsOverride);

  const o = offLuck;
  const d = defLuck;

  return <span ref={topRef}>
      <span>
        <b>Luck Adjustment diagnostics [{name}]</b>
        {showHelp ? <span> <a target="_blank" href="https://hoop-explorer.blogspot.com/2020/07/luck-adjustment-details.html">(?)</a></span> : null}
        {!individualMode ? <span>
          &nbsp;off=[<b>{o.deltaOffAdjEff.toFixed(1)}</b>], def=[<b>{d.deltaDefAdjEff.toFixed(1)}</b>], margin=[<b>{(o.deltaOffAdjEff - d.deltaDefAdjEff).toFixed(1)}</b>] pts/100
         </span> : null}
        &nbsp;(<a href="#" onClick={(event) => { event.preventDefault(); setShowDetails(!showDetails) }}>show {showDetails ? "less" : "more"}</a>)
      </span>
      { showDetails ? <ul>
        <li><b>Offense</b>:
        {!individualMode ? <span> adjustment [<b>{o.deltaOffAdjEff.toFixed(1)}</b>] pts/100</span> : null}</li>
        <ul>
          <li><b>3P Shooting</b>: 3P% [<b>{(100*o.delta3P).toFixed(1)}</b>%], eFG [<b>{(100*o.deltaOffEfg).toFixed(1)}</b>%]
          {!individualMode ? <span>, adjustment [<b>{o.deltaOffAdjEff.toFixed(1)}</b>] pts/100</span> : null }
          &nbsp;(<a href="#" onClick={(event) => { event.preventDefault(); setShow3POff(!show3POff) }}>show {show3POff ? "less" : "more"}</a>)
          </li>
          { show3POff ? <ul>
            <li>Delta_eFG: [<b>{(100*o.deltaOffEfg).toFixed(1)}</b>%] =
            1.5 * (Adj_3P_Off [<b>{(100*(o.delta3P + o.sample3P)).toFixed(1)}</b>%] - Sample_3P_Off [<b>{(100*o.sample3P).toFixed(1)}</b>%]) *
            Sample_Off_3PR [<b>{(100*o.sampleOff3PRate).toFixed(1)}</b>%]
            </li>
            <ul>
              <li><i>(The idea is we'll calculate the expected shooting performance of the players in the lineup set, weighted by their usage, and then regress the actual 3P shooting agains that.)</i></li>
              <li>Adj_3P_Off: [<b>{(100*(o.delta3P + o.sample3P)).toFixed(1)}</b>%] = <i>Weighted_Mean</i>(
                Expected_3P% [<b>{(100*o.sampleBase3P).toFixed(1)}</b>%], Base_3PA [<b>{o.base3PA.toFixed(0)}</b>],
              Sample_3P% [<b>{(100*o.sample3P).toFixed(1)}</b>%], Sample_3PA [<b>{o.sample3PA.toFixed(0)}</b>])</li>
              <li>Expected_3P%: [<b>{(100*o.sampleBase3P).toFixed(1)}</b>%] = Calculated from <i>Weighted_Mean</i> of:</li>
              <ul>
                {_.toPairs(o.player3PInfo).map((pV: [ string, any ], index: number) => {
                  return <li key={index}>[<b>{pV[0]}</b>]: Sample_3PA=[<b>{((pV[1]?.sample3PA || 0)).toFixed(0)}</b>] Base_3P%=[<b>{(100*(pV[1]?.base3P || 0)).toFixed(1)}</b>%]</li>
                })}
              </ul>
            </ul>
            { !individualMode ? // Only show the following if in team mode
            <span>
              <li>Adj_Delta_Pts/100: [<b>{o.deltaOffAdjEff.toFixed(1)}</b>] = Delta_Pts/100 [<b>{o.deltaOffPpp.toFixed(1)}</b>] * D1_Avg_Eff [<b>{o.avgEff.toFixed(1)}</b>] / Sample_Def_SOS [<b>{o.sampleDefSos.toFixed(1)}</b>]</li>
              <li>Delta_Pts/100: [<b>{o.deltaOffPpp.toFixed(1)}</b>] = Delta_Pts_No_ORBs/100 [<b>{o.deltaOffPppNoOrb.toFixed(1)}</b>] + Pts_Off_Delta_Misses/100 [<b>{o.deltaPtsOffMisses.toFixed(1)}</b>]</li>
              <ul>
                <li><i>(Because of the change in 3P%, there are either more or less misses that can be rebounded by the offense.)</i></li>
                <li>Delta_Pts_No_ORBs/100: [<b>{o.deltaOffPppNoOrb.toFixed(1)}</b>] = 2 * Delta_eFG [<b>{(100*o.deltaOffEfg).toFixed(1)}</b>%] * Sample_Off_FGA [<b>{o.sampleOffFGA.toFixed(0)}</b>] / Sample_Possessions [<b>{o.samplePoss.toFixed(0)}</b>]</li>
                <li>Pts_Off_Delta_Misses/100: [<b>{o.deltaPtsOffMisses.toFixed(1)}</b>] = ORB_Factor [<b>{(100*o.deltaOffOrbFactor).toFixed(1)}</b>%] * (Sample_Def_Pts/100 [<b>{o.sampleOffPpp.toFixed(1)}</b>] + Delta_Pts_No_ORBs/100: [<b>{o.deltaOffPppNoOrb.toFixed(1)}</b>])</li>
                <ul>
                  <li>ORB_Factor: [<b>{(100*o.deltaOffOrbFactor).toFixed(1)}</b>%] = Delta_Misses% [<b>{(100*o.deltaMissesPct).toFixed(1)}</b>%] * Sample_Def_ORB [<b>{(100*o.sampleOffOrb).toFixed(1)}</b>%] / (1 - Delta_Misses/100 [<b>{(100*o.deltaMissesPct).toFixed(1)}</b>%] * Sample_Def_ORB [<b>{(100*o.sampleOffOrb).toFixed(1)}</b>%])</li>
                </ul>
              </ul>
            </span> : null }
          </ul> : null }
        </ul>
        <li><b>Defense</b>:
          {!individualMode ? <span> adjustment [<b>{d.deltaDefAdjEff.toFixed(1)}</b>] pts/100</span> : null }
        </li>
        <ul>
          <li><b>3P Shooting</b>: 3P% [<b>{(100*d.delta3P).toFixed(1)}</b>%]
            {!individualMode ? <span>, eFG [<b>{(100*d.deltaDefEfg).toFixed(1)}</b>%], adjustment [<b>{d.deltaDefAdjEff.toFixed(1)}</b>] pts/100</span> : null}
           &nbsp;(<a href="#" onClick={(event) => { event.preventDefault(); setShow3PDef(!show3PDef) }}>show {show3PDef ? "less" : "more"}</a>)
          </li>
          { show3PDef ? <ul>
            {individualMode ?
              <li>Delta_3P%: [<b>{(100*d.delta3P).toFixed(1)}</b>%] =
            Adj_3P_Def [<b>{(100*d.adjDef3P).toFixed(1)}</b>%] - Sample_3P_Def [<b>{(100*d.sampleDef3P).toFixed(1)}</b>%]</li>
              : <li>Delta_eFG: [<b>{(100*d.deltaDefEfg).toFixed(1)}</b>%] =
            1.5 * (Adj_3P_Def [<b>{(100*d.adjDef3P).toFixed(1)}</b>%] - Sample_3P_Def [<b>{(100*d.sampleDef3P).toFixed(1)}</b>%]) *
            Sample_Def_3PR [<b>{(100*d.sampleDef3PRate).toFixed(1)}</b>%]</li>
            }
            <ul>
              <li><i>(The idea is we calculate a "3P defense" number from the sample, regressed to the [<b>{baseline}</b>], assuming that [<b>{(100*d.luckPct).toFixed(1)}</b>%] of it is just luck.
              Then we calculate a "luck adjusted" 3P% by combining "3P defense" and the weighted average opponent 3P%.)
              </i></li>
              <li>Adj_3P_Def: [<b>{(100*d.adjDef3P).toFixed(1)}</b>%] = Sample_3P_SoS [<b>{(100*d.sampleDef3PSos).toFixed(1)}</b>%] + Delta_3PD [<b>{(100*d.avg3PSosAdj).toFixed(1)}</b>%]</li>
              <li>Delta_3PD: [<b>{(100*d.avg3PSosAdj).toFixed(1)}</b>%] = <i>Weighted_Mean</i>(
                Base_Delta_3PD [<b>{(100*d.base3PSosAdj).toFixed(1)}</b>%], Base_Possessions [<b>{d.basePoss.toFixed(0)}</b>],
              Sample_Delta_3PD [<b>{(100*d.sample3PSosAdj).toFixed(1)}</b>%], Sample_Possessions [<b>{d.samplePoss.toFixed(0)}</b>])</li>
              <ul>
                <li>Base_Delta_3PD: [<b>{(100*d.base3PSosAdj).toFixed(1)}</b>%] = (1 - Luck_Factor [<b>{(100*d.luckPct).toFixed(1)}</b>%]) *
                (Base_3P_Def [<b>{(100*d.baseDef3P).toFixed(1)}</b>%] - Base_3P_SoS [<b>{(100*d.baseDef3PSos).toFixed(1)}</b>%])
                </li>
                <li>Sample_Delta_3PD: [<b>{(100*d.sample3PSosAdj).toFixed(1)}</b>%] = (1 - Luck_Factor [<b>{(100*d.luckPct).toFixed(1)}</b>%]) *
                (Sample_3P_Def [<b>{(100*d.sampleDef3P).toFixed(1)}</b>%] - Sample_3P_SoS [<b>{(100*d.sampleDef3PSos).toFixed(1)}</b>%])
                </li>
              </ul>
            </ul>
            {!individualMode ? <li>Adj_Delta_Pts/100: [<b>{d.deltaDefAdjEff.toFixed(1)}</b>] = Delta_Pts/100 [<b>{d.deltaDefPpp.toFixed(1)}</b>] * D1_Avg_Eff [<b>{d.avgEff.toFixed(1)}</b>] / Sample_Off_SOS [<b>{d.sampleOffSos.toFixed(1)}</b>]</li> : null}
            {!individualMode ? <li>Delta_Pts/100: [<b>{d.deltaDefPpp.toFixed(1)}</b>] = Delta_Pts_No_ORBs/100 [<b>{d.deltaDefPppNoOrb.toFixed(1)}</b>] + Pts_Off_Delta_Misses/100 [<b>{d.deltaPtsOffMisses.toFixed(1)}</b>]</li> : null}
            {!individualMode ? <ul>
              <li><i>(Because of the change in 3P%, there are either more or less misses that can be rebounded by the offense.)</i></li>
              <li>Delta_Pts_No_ORBs/100: [<b>{d.deltaDefPppNoOrb.toFixed(1)}</b>] = 2 * Delta_eFG [<b>{(100*d.deltaDefEfg).toFixed(1)}</b>%] * Sample_Def_FGA [<b>{d.sampleDefFGA.toFixed(0)}</b>] / Sample_Possessions [<b>{d.samplePoss.toFixed(0)}</b>]</li>
              <li>Pts_Off_Delta_Misses/100: [<b>{d.deltaPtsOffMisses.toFixed(1)}</b>] = ORB_Factor [<b>{(100*d.deltaDefOrbFactor).toFixed(1)}</b>%] * (Sample_Def_Pts/100 [<b>{d.sampleDefPpp.toFixed(1)}</b>] + Delta_Pts_No_ORBs/100: [<b>{d.deltaDefPppNoOrb.toFixed(1)}</b>])</li>
              <ul>
                <li>ORB_Factor: [<b>{(100*d.deltaDefOrbFactor).toFixed(1)}</b>%] = Delta_Misses% [<b>{(100*d.deltaMissesPct).toFixed(1)}</b>%] * Sample_Def_ORB [<b>{(100*d.sampleDefOrb).toFixed(1)}</b>%] / (1 - Delta_Misses/100 [<b>{(100*d.deltaMissesPct).toFixed(1)}</b>%] * Sample_Def_ORB [<b>{(100*d.sampleDefOrb).toFixed(1)}</b>%])</li>
              </ul>
            </ul> : null}
          </ul> : null }
        </ul>
        {baseline == "baseline" ? <li><i>(Regressing over baseline uses in a smaller and possibly biased dataset, which is bad; but can be useful if there is reason to believe the characteristics
        of the team are different compared to the sample - eg due to injury.)</i></li> : null}
      </ul> : null }
    </span>;
};
export default LuckAdjDiagView;
