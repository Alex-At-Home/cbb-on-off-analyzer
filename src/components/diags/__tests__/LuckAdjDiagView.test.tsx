import renderer from 'react-test-renderer';
import React from 'react';
import LuckAdjDiagView from '../LuckAdjDiagView';
import { sampleOffOnOffLuckDiagnostics, sampleDefOnOffLuckDiagnostics } from "../../../sample-data/sampleOnOffLuckDiagnostics";
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe("LuckAdjDiagView", () => {
  test("LuckAdjDiagView - should create snapshot (season, !override)", () => {

    const wrapper = shallow(
      <LuckAdjDiagView
        name="Test1"
        offLuck={sampleOffOnOffLuckDiagnostics}
        defLuck={sampleDefOnOffLuckDiagnostics}
        baseline="season"
        showHelp={false}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test("LuckAdjDiagView - should create snapshot (baseline, override)", () => {

    const wrapper = shallow(
      <LuckAdjDiagView
        name="Test2"
        offLuck={sampleOffOnOffLuckDiagnostics}
        defLuck={sampleDefOnOffLuckDiagnostics}
        baseline="baseline"
        showHelp={true}
        showDetailsOverride={true}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
