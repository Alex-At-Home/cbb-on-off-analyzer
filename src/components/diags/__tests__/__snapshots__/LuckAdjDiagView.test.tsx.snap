// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuckAdjDiagView LuckAdjDiagView - should create snapshot (baseline, override) 1`] = `
<span>
  <span>
    <b>
      Luck Adjustment diagnostics [
      Test2
      ]
    </b>
    <span>
       
      <a
        href="https://hoop-explorer.blogspot.com/2020/07/luck-adjustment-details.html"
        target="_blank"
      >
        (?)
      </a>
    </span>
     off=[
    <b>
      -1.8
    </b>
    ], def=[
    <b>
      7.7
    </b>
    ], margin=[
    <b>
      -9.5
    </b>
    ] pts/100  (
    <a
      href="#"
      onClick={[Function]}
    >
      show 
      less
    </a>
    )
  </span>
  <ul>
    <li>
      <b>
        Offense
      </b>
      : adjustment [
      <b>
        -1.8
      </b>
      ] pts/100
    </li>
    <ul>
      <li>
        <b>
          3P Shooting
        </b>
        : 3P% [
        <b>
          -2.2
        </b>
        %], eFG [
        <b>
          -1.2
        </b>
        %], adjustment [
        <b>
          -1.8
        </b>
        ] pts/100 (
        <a
          href="#"
          onClick={[Function]}
        >
          show 
          less
        </a>
        )
      </li>
      <ul>
        <li>
          Delta_eFG: [
          <b>
            -1.2
          </b>
          %] = 1.5 * (Adj_3P_Off [
          <b>
            33.0
          </b>
          %] - Sample_3P_Off [
          <b>
            35.3
          </b>
          %]) * Sample_Off_3PR [
          <b>
            37.3
          </b>
          %]
        </li>
        <ul>
          <li>
            <i>
              (The idea is we'll calculate the expected shooting performance of the players in the lineup set, weighted by their usage, and then regress the actual 3P shooting agains that.)
            </i>
          </li>
          <li>
            Adj_3P_Off: [
            <b>
              33.0
            </b>
            %] = 
            <i>
              Weighted_Mean
            </i>
            ( Expected_3P% [
            <b>
              31.7
            </b>
            %], Base_3PA [
            <b>
              800
            </b>
            ], Sample_3P% [
            <b>
              35.3
            </b>
            %], Sample_3PA [
            <b>
              500
            </b>
            ])
          </li>
          <li>
            Expected_3P%: [
            <b>
              31.7
            </b>
            %] = Calculated from 
            <i>
              Weighted_Mean
            </i>
             of:
          </li>
          <ul>
            <li
              key="0"
            >
              [
              <b>
                Cowan, Anthony
              </b>
              ]: Sample_3PA=[
              <b>
                122
              </b>
              ] Base_3P%=[
              <b>
                32.2
              </b>
              %]
            </li>
            <li
              key="1"
            >
              [
              <b>
                Wiggins, Aaron
              </b>
              ]: Sample_3PA=[
              <b>
                163
              </b>
              ] Base_3P%=[
              <b>
                31.3
              </b>
              %]
            </li>
          </ul>
        </ul>
        <li>
          Adj_Delta_Pts/100: [
          <b>
            -1.8
          </b>
          ] = Delta_Pts/100 [
          <b>
            -1.8
          </b>
          ] * D1_Avg_Eff [
          <b>
            100.0
          </b>
          ] / Sample_Def_SOS [
          <b>
            98.3
          </b>
          ]
        </li>
        <li>
          Delta_Pts/100: [
          <b>
            -1.8
          </b>
          ] = Delta_Pts_No_ORBs/100 [
          <b>
            -2.1
          </b>
          ] + Pts_Off_Delta_Misses/100 [
          <b>
            0.3
          </b>
          ]
        </li>
        <ul>
          <li>
            <i>
              (Because of the change in 3P%, there are either more or less misses that can be rebounded by the offense.)
            </i>
          </li>
          <li>
            Delta_Pts_No_ORBs/100: [
            <b>
              -2.1
            </b>
            ] = 2 * Delta_eFG [
            <b>
              -1.2
            </b>
            %] * Sample_Off_FGA [
            <b>
              1100
            </b>
            ] / Sample_Possessions [
            <b>
              1293
            </b>
            ]
          </li>
          <li>
            Pts_Off_Delta_Misses/100: [
            <b>
              0.3
            </b>
            ] = ORB_Factor [
            <b>
              0.3
            </b>
            %] * (Sample_Def_Pts/100 [
            <b>
              111.4
            </b>
            ] + Delta_Pts_No_ORBs/100: [
            <b>
              -2.1
            </b>
            ])
          </li>
          <ul>
            <li>
              ORB_Factor: [
              <b>
                0.3
              </b>
              %] = Delta_Misses% [
              <b>
                0.8
              </b>
              %] * Sample_Def_ORB [
              <b>
                33.7
              </b>
              %] / (1 - Delta_Misses/100 [
              <b>
                0.8
              </b>
              %] * Sample_Def_ORB [
              <b>
                33.7
              </b>
              %])
            </li>
          </ul>
        </ul>
      </ul>
    </ul>
    <li>
      <b>
        Defense
      </b>
      : adjustment [
      <b>
        7.7
      </b>
      ] pts/100
    </li>
    <ul>
      <li>
        <b>
          3P Shooting
        </b>
        : 3P% [
        <b>
          6.1
        </b>
        %], eFG [
        <b>
          3.4
        </b>
        %], adjustment [
        <b>
          7.7
        </b>
        ] pts/100 (
        <a
          href="#"
          onClick={[Function]}
        >
          show 
          less
        </a>
        )
      </li>
      <ul>
        <li>
          Delta_eFG: [
          <b>
            3.4
          </b>
          %] = 1.5 * (Adj_3P_Def [
          <b>
            36.3
          </b>
          %] - Sample_3P_Def [
          <b>
            30.2
          </b>
          %]) * Sample_Def_3PR [
          <b>
            36.9
          </b>
          %]
        </li>
        <ul>
          <li>
            <i>
              (The idea is we calculate a "3P defense" number from the sample, regressed to the [
              <b>
                baseline
              </b>
              ], assuming that [
              <b>
                66.0
              </b>
              %] of it is just luck. Then we calculate a "luck adjusted" 3P% by combining "3P defense" and the weighted average opponent 3P%.)
            </i>
          </li>
          <li>
            Adj_3P_Def: [
            <b>
              36.3
            </b>
            %] = Sample_3P_SoS [
            <b>
              37.2
            </b>
            %] + Delta_3PD [
            <b>
              -1.0
            </b>
            %]
          </li>
          <li>
            Delta_3PD: [
            <b>
              -1.0
            </b>
            %] = 
            <i>
              Weighted_Mean
            </i>
            ( Base_Delta_3PD [
            <b>
              -0.5
            </b>
            %], Base_Possessions [
            <b>
              2089
            </b>
            ], Sample_Delta_3PD [
            <b>
              -2.4
            </b>
            %], Sample_Possessions [
            <b>
              596
            </b>
            ])
          </li>
          <ul>
            <li>
              Base_Delta_3PD: [
              <b>
                -0.5
              </b>
              %] = (1 - Luck_Factor [
              <b>
                66.0
              </b>
              %]) * (Base_3P_Def [
              <b>
                34.0
              </b>
              %] - Base_3P_SoS [
              <b>
                35.6
              </b>
              %])
            </li>
            <li>
              Sample_Delta_3PD: [
              <b>
                -2.4
              </b>
              %] = (1 - Luck_Factor [
              <b>
                66.0
              </b>
              %]) * (Sample_3P_Def [
              <b>
                30.2
              </b>
              %] - Sample_3P_SoS [
              <b>
                37.2
              </b>
              %])
            </li>
          </ul>
        </ul>
        <li>
          Adj_Delta_Pts/100: [
          <b>
            7.7
          </b>
          ] = Delta_Pts/100 [
          <b>
            8.4
          </b>
          ] * D1_Avg_Eff [
          <b>
            100.0
          </b>
          ] / Sample_Off_SOS [
          <b>
            109.2
          </b>
          ]
        </li>
        <li>
          Delta_Pts/100: [
          <b>
            8.4
          </b>
          ] = Delta_Pts_No_ORBs/100 [
          <b>
            9.0
          </b>
          ] + Pts_Off_Delta_Misses/100 [
          <b>
            -0.6
          </b>
          ]
        </li>
        <ul>
          <li>
            <i>
              (Because of the change in 3P%, there are either more or less misses that can be rebounded by the offense.)
            </i>
          </li>
          <li>
            Delta_Pts_No_ORBs/100: [
            <b>
              9.0
            </b>
            ] = 2 * Delta_eFG [
            <b>
              3.4
            </b>
            %] * Sample_Def_FGA [
            <b>
              795
            </b>
            ] / Sample_Possessions [
            <b>
              596
            </b>
            ]
          </li>
          <li>
            Pts_Off_Delta_Misses/100: [
            <b>
              -0.6
            </b>
            ] = ORB_Factor [
            <b>
              -0.6
            </b>
            %] * (Sample_Def_Pts/100 [
            <b>
              91.6
            </b>
            ] + Delta_Pts_No_ORBs/100: [
            <b>
              9.0
            </b>
            ])
          </li>
          <ul>
            <li>
              ORB_Factor: [
              <b>
                -0.6
              </b>
              %] = Delta_Misses% [
              <b>
                -2.2
              </b>
              %] * Sample_Def_ORB [
              <b>
                25.8
              </b>
              %] / (1 - Delta_Misses/100 [
              <b>
                -2.2
              </b>
              %] * Sample_Def_ORB [
              <b>
                25.8
              </b>
              %])
            </li>
          </ul>
        </ul>
      </ul>
    </ul>
    <li>
      <i>
        (Regressing over baseline uses in a smaller and possibly biased dataset, which is bad; but can be useful if there is reason to believe the characteristics of the team are different compared to the sample - eg due to injury.)
      </i>
    </li>
  </ul>
</span>
`;

exports[`LuckAdjDiagView LuckAdjDiagView - should create snapshot (season, !override) 1`] = `
<span>
  <span>
    <b>
      Luck Adjustment diagnostics [
      Test1
      ]
    </b>
     off=[
    <b>
      -1.8
    </b>
    ], def=[
    <b>
      7.7
    </b>
    ], margin=[
    <b>
      -9.5
    </b>
    ] pts/100  (
    <a
      href="#"
      onClick={[Function]}
    >
      show 
      less
    </a>
    )
  </span>
  <ul>
    <li>
      <b>
        Offense
      </b>
      : adjustment [
      <b>
        -1.8
      </b>
      ] pts/100
    </li>
    <ul>
      <li>
        <b>
          3P Shooting
        </b>
        : 3P% [
        <b>
          -2.2
        </b>
        %], eFG [
        <b>
          -1.2
        </b>
        %], adjustment [
        <b>
          -1.8
        </b>
        ] pts/100 (
        <a
          href="#"
          onClick={[Function]}
        >
          show 
          more
        </a>
        )
      </li>
    </ul>
    <li>
      <b>
        Defense
      </b>
      : adjustment [
      <b>
        7.7
      </b>
      ] pts/100
    </li>
    <ul>
      <li>
        <b>
          3P Shooting
        </b>
        : 3P% [
        <b>
          6.1
        </b>
        %], eFG [
        <b>
          3.4
        </b>
        %], adjustment [
        <b>
          7.7
        </b>
        ] pts/100 (
        <a
          href="#"
          onClick={[Function]}
        >
          show 
          more
        </a>
        )
      </li>
    </ul>
  </ul>
</span>
`;
